### PR TITLE
[`FA2`] Add flash attention for opt

### PR DESCRIPTION
**Pull request by susnato**
# What does this PR do?

Adds Flash Attention 2 for `OPT` as discussed in in this issue - #26350 .

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc : @younesbelkada

### CODE REVIEW

**Comment by susnato** [comment]
The tests were done using a RTX 3060 (Ampere) which supports Flash Attention 2.
Screenshot from 2023-09-26 16-23-55

**Review by younesbelkada — COMMENTED** [review]
The changes look great to me! Thanks a lot for working on this @susnato !
I can confirm the tests pass on a A100, just tested it
Screenshot 2023-09-26 at 13 00 02

Before merging please have a look at my comment below, also can you add OPT to the list of officially supported models? The changes would go here: https://github.com/susnato/transformers/blob/flash_attn_opt/docs/source/en/perf_infer_gpu_one.md

**Review by younesbelkada — COMMENTED** [review]
@susnato I have realised there were some issues with FA 2 + OPT + use_cache, some minor changes were needed, please have a look at https://github.com/susnato/transformers/pull/3 and I will add later tests for the case users use caching to catch that in the future. 

To reproduce: 
```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

model = AutoModelForCausalLM.from_pretrained(
    "facebook/opt-350m",
    use_flash_attention_2=True,
    torch_dtype=torch.float16
).to(0)

tokenizer = AutoTokenizer.from_pretrained("facebook/opt-350m")

text = "Hello my name is"
inputs = tokenizer(text, return_tensors="pt").to(0)

out = model.generate(**inputs, max_new_tokens=30, use_cache=True)
print(tokenizer.batch_decode(out, skip_special_tokens=True))
```

**Comment by susnato** [comment]
Hi @younesbelkada, I have added the model to the list. Also I have merged the [susnato#3](https://github.com/susnato/transformers/pull/3), please let me know if anymore changes are needed or not.

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for iterating ! Can you also run the styling checks? 
```bash
make fixup
```

**Comment by susnato** [comment]
Hi @younesbelkada , done.

File: tests/models/opt/test_modeling_opt.py
Comment (younesbelkada) [inline tests/models/opt/test_modeling_opt.py:219]: I think this should not be removed

**Review by younesbelkada — COMMENTED** [review]
Thanks for iterating! left one comment

File: tests/models/opt/test_modeling_opt.py
Comment (susnato) [reply tests/models/opt/test_modeling_opt.py:219]: It seems to get error for the fx tests - `test_torch_fx` and `test_torch_fx_output_loss`.
The error is happening for this [line](https://github.com/susnato/transformers/blob/74e968760c140fde0d05caad66d52ddbcb911a30/src/transformers/models/opt/modeling_opt.py#L872).

Comment (younesbelkada) [reply tests/models/opt/test_modeling_opt.py:219]: thanks, I see, in that line can you use instead `torch.isin` ? https://pytorch.org/docs/stable/generated/torch.isin.html - suggestion from @michaelbenayoun

File: src/transformers/models/opt/modeling_opt.py
Comment (younesbelkada) [inline src/transformers/models/opt/modeling_opt.py:872]: Can you try to add a conditional check to check if there is any 0 in the attention mask, sometimes users pass an attention mask with a full ones, therefore we'll need to set `padding_mask` to None to avoid entering the padding case in `OPTFlashAttention2` module

If the test fails with torch.fx can you try with `torch.isin` ? otherwise you can also do a check based on the sum of the attention mask

Comment (susnato) [reply src/transformers/models/opt/modeling_opt.py:872]: Ok, sorry for removing it directly without discussing.

Comment (younesbelkada) [reply src/transformers/models/opt/modeling_opt.py:872]: no problem at all @susnato ! 🙏

Comment (susnato) [reply src/transformers/models/opt/modeling_opt.py:872]: It fails for - 
- `torch.isin(attention_mask.int(), 0).sum().item()!=0`(AssertionError: Couldn't trace module: symbolically traced variables cannot be used as inputs to control flow
) 
- `torch.isin(attention_mask.int(), 0).sum()!=0`(AssertionError: Couldn't trace module: __bool__ should return bool, returned Tensor
)
- `attention_mask.sum().item()==attention_mask.numel()`(AssertionError: Couldn't trace module: symbolically traced variables cannot be used as inputs to control flow
)
- `if not torch.equal(attention_mask, torch.ones_like(attention_mask))`( AssertionError: Couldn't trace module: symbolically traced variables cannot be used as inputs to control flow)

Comment (susnato) [reply src/transformers/models/opt/modeling_opt.py:872]: There seems to be a related open issue [here](https://github.com/pytorch/pytorch/issues/54602)

File: src/transformers/models/opt/modeling_opt.py
Comment (susnato) [inline src/transformers/models/opt/modeling_opt.py:418]: It seems that if we check the ones array here the `torch.fx` tests pass.

File: src/transformers/models/opt/modeling_opt.py
Comment (susnato) [inline src/transformers/models/opt/modeling_opt.py:873]: In that case we need to pass `padding_mask = attention_mask` and we check for the ones array in the `_flash_attention_forward`. If we use any `if-else` condition here it gives an error for the `torch.fx` tests.

**Review by susnato — COMMENTED** [review]
Hi @younesbelkada, I have solved the error but it requires us to change the `_flash_attention_forward` function a little bit, otherwise if we try to add the `if` statement in the `OPTDecoder.forward` we are getting the error.

**Review by younesbelkada — COMMENTED** [review]
Thanks very much for iterating @susnato I have tried out also different approaches on my end as well

```python
            # Avoids torch FX issues
            if torch.any(~attention_mask.bool()):
                padding_mask = attention_mask
            else:
                padding_mask = None
```
and the FX tests still fails 

The issue with the current approach is that the `padding_mask.sum().item() != padding_mask.numel()` adds an important overhead as it computed by each attention layer, especially for long sequence length leading to a slow down. 

Let me dig a bit with @michaelbenayoun regarding failing FX tests and get back to you

**Comment by susnato** [comment]
Thanks for the reply @younesbelkada, in meantime I would like to work on adding flash attention for another model.

**Comment by younesbelkada** [comment]
Thanks @susnato , sure yes ok, let me know on #26350 which model would you take, perhaps you can try your hands on Starcoder if you are interested. Let me know!

**Comment by susnato** [comment]
yes I am interested in that! @younesbelkada

**Comment by younesbelkada** [comment]
Thanks very much ! Looking forward to your PR !

**Comment by susnato** [comment]
Also please assign me in the list :). @younesbelkada

**Comment by dathudeptrai** [comment]
@susnato @younesbelkada why not use `F.scaled_dot_product_attention` instead as it support Flash-Attention-2 and able to use with `torch.compile` ?. I used to use the `flash-attention` on official repo like this PR but now I moved to the `F.scaled_dot_product_attention`.

**Comment by younesbelkada** [comment]
Hi @susnato 
Thanks for your patience,
I had a look at the issue with @michaelbenayoun and the solution is the following: 

1- define a method on the top level of the modeling opt file that checks if there is any padding token inside the attention mask
```python

if is_torch_fx_available():
    @torch.fx.wrap
    def check_padding_in_attention_mask(attention_mask):
        if 0 in attention_mask:
            return attention_mask
        return None
else:
    def check_padding_in_attention_mask(attention_mask):
        if 0 in attention_mask:
            return attention_mask
        return None
```
And add the `@torch.fx.wrap` decorator in the method to make sure it is compatible with the PT versions we support. 
2- revert until the commit [689f599](https://github.com/huggingface/transformers/pull/26414/commits/689f599132f121a3716cf4497570cbbda5c01737) and replace the simple logic by a call to that method 

That way FX tests should hopefully pass

I will later take care of moving `check_padding_in_attention_mask` in `pytorch_utils` file so that models that support FX tracing can use that method for future integrations.

**Comment by susnato** [comment]
Hi @dathudeptrai, `transformers` maintain a single modeling file policy, where we want to make sure that someone who is reading the code is properly able to understand the modeling code without having to switch to any other file. That's why even the commonly used `dot product attention` which is used by many modes are defined for each one. We define `F.scaled_dot_product_attention` so that the user does not have to visit the torch website to understand the implementation. 

I believe that in future when flash attention becomes more stable the maintainers will try to implement the function within the library itself(given that it's possible) to reduce the dependency. 

Also @younesbelkada could give you a better explanation and correct me if I am wrong here :).

**Comment by susnato** [comment]
Hi @younesbelkada, I have added the solution that you suggested and it's working!
Let me know if any more changes are required or not.

File: src/transformers/models/opt/modeling_opt.py
Comment (younesbelkada) [inline src/transformers/models/opt/modeling_opt.py:87]: I will properly docstring this and move it inside `pytorch_utils` as it will be used by all models that will support FA2 and already supports fx tracing. I will take care of that in a separate PR.
```suggestion
# TODO: @younesbelkada move that in pytorch_utils and document it
if is_torch_fx_available():

    @torch.fx.wrap
    def check_padding_in_attention_mask(attention_mask):
        if 0 in attention_mask:
            return attention_mask
        return None

else:

    def check_padding_in_attention_mask(attention_mask):
        if 0 in attention_mask:
            return attention_mask
        return None
```

**Review by younesbelkada — COMMENTED** [review]
Great addition thank you very much !
Can you confirm all slow tests pass ? (you need to merge with main as I added a new FA-2 test) https://github.com/huggingface/transformers/pull/26415 
Then we can merge I think
Thanks!

**Comment by younesbelkada** [comment]
@dathudeptrai Thanks for your question
We are currently thinking of migrating SDPA natively into transformers in the next weeks, we will keep you posted with @fxmarty

**Comment by susnato** [comment]
Hi @younesbelkada the `test_flash_attn_2_generate_use_cache` is failing for both `llama` and `opt` models :confused: .
Screenshot from 2023-09-28 15-52-25

Screenshot from 2023-09-28 15-51-01

Trying to figure out the reason.

**Comment by susnato** [comment]
@younesbelkada I already did :(.

**Comment by younesbelkada** [comment]
@susnato can you share the full traceback?

**Comment by susnato** [comment]
<details>
<summary>Full Traceback for Llama</summary>


self = <tests.models.llama.test_modeling_llama.LlamaModelTest testMethod=test_flash_attn_2_generate_use_cache>

    @require_flash_attn
    @require_torch_gpu
    @mark.flash_attn_test
    @slow
    def test_flash_attn_2_generate_use_cache(self):
        import torch
    
        for model_class in self.all_generative_model_classes:
            if not model_class._supports_flash_attn_2:
                return
    
            config, _ = self.model_tester.prepare_config_and_inputs_for_common()
            model = model_class(config)
    
            with tempfile.TemporaryDirectory() as tmpdirname:
                model.save_pretrained(tmpdirname)
    
                dummy_input = torch.LongTensor([[0, 2, 3, 4], [0, 2, 3, 4]]).to(torch_device)
                dummy_attention_mask = torch.LongTensor([[1, 1, 1, 1], [1, 1, 1, 0]]).to(torch_device)
    
                model = model_class.from_pretrained(
                    tmpdirname, torch_dtype=torch.float16, use_flash_attention_2=True, low_cpu_mem_usage=True
                ).to(torch_device)
    
                # Just test that a large cache works as expected
>               _ = model.generate(
                    dummy_input, attention_mask=dummy_attention_mask, max_new_tokens=30, do_sample=False
                )

tests/test_modeling_common.py:2936: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../anaconda3/envs/transformers/lib/python3.9/site-packages/torch/autograd/grad_mode.py:27: in decorate_context
    return func(*args, **kwargs)
src/transformers/generation/utils.py:1606: in generate
    return self.greedy_search(
src/transformers/generation/utils.py:2454: in greedy_search
    outputs = self(
../../anaconda3/envs/transformers/lib/python3.9/site-packages/torch/nn/modules/module.py:1194: in _call_impl
    return forward_call(*input, **kwargs)
src/transformers/models/llama/modeling_llama.py:1034: in forward
    outputs = self.model(
../../anaconda3/envs/transformers/lib/python3.9/site-packages/torch/nn/modules/module.py:1194: in _call_impl
    return forward_call(*input, **kwargs)
src/transformers/models/llama/modeling_llama.py:921: in forward
    layer_outputs = decoder_layer(
../../anaconda3/envs/transformers/lib/python3.9/site-packages/torch/nn/modules/module.py:1194: in _call_impl
    return forward_call(*input, **kwargs)
src/transformers/models/llama/modeling_llama.py:631: in forward
    hidden_states, self_attn_weights, present_key_value = self.self_attn(
../../anaconda3/envs/transformers/lib/python3.9/site-packages/torch/nn/modules/module.py:1194: in _call_impl
    return forward_call(*input, **kwargs)
src/transformers/models/llama/modeling_llama.py:489: in forward
    attn_output = self._flash_attention_forward(
src/transformers/models/llama/modeling_llama.py:546: in _flash_attention_forward
    attn_output = pad_input(attn_output_unpad, indices_q, batch_size, query_length)
../../anaconda3/envs/transformers/lib/python3.9/site-packages/flash_attn/bert_padding.py:208: in pad_input
    output = index_put_first_axis(hidden_states, indices, batch * seqlen)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

ctx = <torch.autograd.function.IndexPutFirstAxisBackward object at 0x7f1fb51b9040>
values = tensor([[[-2.3666e-02,  2.1423e-02, -7.4646e-02,  3.6530e-02, -7.8979e-02,
          -3.4546e-02,  1.0236e-01,  4.3549...4504e-03, -9.3384e-02,
          -4.5532e-02, -5.5847e-02,  4.0253e-02]]], device='cuda:0',
       dtype=torch.float16)
indices = tensor([0, 1], device='cuda:0', dtype=torch.int32), first_axis_dim = 2

    @staticmethod
    def forward(ctx, values, indices, first_axis_dim):
        ctx.save_for_backward(indices)
        assert indices.ndim == 1
        assert values.ndim >= 2
        output = torch.zeros(
            first_axis_dim, *values.shape[1:], device=values.device, dtype=values.dtype
        )
        # TD [2022-03-04] For some reason torch.scatter is a bit faster than indexing.
>       output[indices] = values
E       IndexError: tensors used as indices must be long, byte or bool tensors

../../anaconda3/envs/transformers/lib/python3.9/site-packages/flash_attn/bert_padding.py:51: IndexError


</details>

BTW I updated the flash attention library and merged to main but the error is still there.

File: src/transformers/models/opt/modeling_opt.py
Comment (susnato) [reply src/transformers/models/opt/modeling_opt.py:88]: done

**Comment by susnato** [comment]
If you don't mind @younesbelkada , could you please checkout this branch and run the tests on your end and let me know the results?
If this passes on your machine, then It could be due to some unknown error on my end.

**Comment by younesbelkada** [comment]
Sure @susnato no problem, will run the tests locally tomorrow and let you know how it goes

**Comment by susnato** [comment]
Hi @younesbelkada, did you manage to check if the tests run successfully on your local machine?

**Comment by younesbelkada** [comment]
Yes the tests seemed to pass the time I ran them, however since #26792 being merged you need to remove the support for `padding_mask` and follow whqat has been done in that PR. Let me know if you need help on this!

**Comment by susnato** [comment]
After I finish updating the `gpt_bigcode` I will take this up.

**Comment by younesbelkada** [comment]
OK awesome, thanks @susnato !

**Comment by susnato** [comment]
Hi @younesbelkada, I have pushed the changes.
All the flash attention tests are running fine except `test_flash_attn_2_generate_use_cache` but I believe it's due to some problem with my local CUDA installation since the model generates outputs quite well when I run separately.

Please review it and let me know if any more changes are needed.

BTW the CI is failing on `check_repository_consistency` which seems to be unrelated to this PR.

**Comment by susnato** [comment]
The CI is green now!

**Comment by susnato** [comment]
Hey @younesbelkada , are you going to check the speed-up for this model too?  

I am asking this out of curiosity - Are you going to check speed-ups for every model, FlashAttention is added to or is this benchmark needed for popular ones only?

File: src/transformers/models/opt/modeling_opt.py
Comment (younesbelkada) [inline src/transformers/models/opt/modeling_opt.py:117]: Are we ok modifying the Attention Class's signature? We've had issues in the past when changing signature of such classes (e.g. for Llama when we added `padding_mask`) cc @amyeroberts

**Review by younesbelkada — COMMENTED** [review]
Nice thanks a lot for your huge work ! 
I can confirm interesting speedups for opt-350m and opt-2.7b 

opt-2.7b + large seqlen:

Screenshot 2023-11-07 at 16 55 16

opt-350m + small seqlen:

Screenshot 2023-11-07 at 16 55 38

I just left a single question about changing the attention class' init signature, otherwise LGTM !

File: src/transformers/models/opt/modeling_opt.py
Comment (amyeroberts) [reply src/transformers/models/opt/modeling_opt.py:117]: It should be OK, as this isn't a class importable from the top layer of transformers or in the documentation. However, it definitely can still cause issues! What I'd suggest is something like this (happy for you to modify as desired): 

```py
    def __init__(
        self,
        config,
        is_decoder: bool = False,
        **kwargs
    ):
	def _handle_deprecated_argument(config_arg_name, config, fn_arg_name, kwargs):
	    """
	    If a the deprecated argument `fn_arg_name` is passed, raise a deprecation 
	    warning and return that value, otherwise take the equivalent config.config_arg_name
	    """
	    val = None
	    if fn_arg_name in kwargs:
	    	logging.warning(
	       		"Passing in {} to {self.__class__.__name__} is deprecated and won't be supported from v4.38."
	       		" Please set it in the config instead"
	       	)
	       	val = kwargs.pop(fn_arg_name)
	    else:
	       	val = getattr(config, config_arg_name)
            return val

        embed_dim = _handle_deprecated_argument("hidden_size", config, "embed_dim", kwargs)
        num_heads = _handle_deprecated_argument("num_attention_heads", config, "num_heads", kwargs)
        dropout = _handle_deprecated_argument("attention_dropout", config, "dropout", kwargs)
        enable_bias = _handle_deprecated_argument("enable_bias", config, "bias", kwargs)
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! 

Just some comments on handling old arguments + deprecation warnings

**Comment by susnato** [comment]
Hi @amyeroberts, I have pushed the changes you suggested, in addition to that I have also updated the model_doc file and added speedup graphs.

**Review by amyeroberts — APPROVED** [review]
@susnato Thanks for iterating! All looks good to me. 

Let's get a second review from @younesbelkada before merging as he's more familiar with the Flash Attention code

**Review by younesbelkada — APPROVED** [review]
Thanks for your great work ! Can you just confirm the slow integration tests for OPT pass before merging? 🙏

**Comment by susnato** [comment]
Yes, I just checked and all tests are passing! @younesbelkada!

Screenshot from 2023-11-23 01-04-43

**Review by younesbelkada — APPROVED** [review]
Thanks! I'll let @amyeroberts merge the PR

**Pull request merged**
